Improve cilium getEndpointList function

fristonio opened the pull request:
* Currently when doing `cilium endpoint list` a number of goroutines equal to the number of endpoints are created which then extract the current state of individual endpoints. This does not make much sense since we don't have any blocking work when getting model for an endpoint using `endpoint.GetModel`.

* This pull request creates a pool of goroutines with a limit on the maximum number of goroutines(equal to the number of CPUs) and then distribute this work among those.

We ran a bunch of benchmarks to test memory spikes when using different number of concurrent goroutines. It turns out that the heap memory does not show spikes even with a large number of goroutines. This PR thus introduces changes to have a maximum goroutine pool equal to `NumCPU`, which does not alter the performance at all. The benchmark was performed using 50 endpoints in the following three settings -

* Current master - One goroutine for each endpoint.
* Patched with a goroutine pool distributing work for all the endpoints.
* Cilium v1.5

The pprof output for each of these cases are attached below for reference.
[PProf Output](https://drive.google.com/open?id=1B_96_6CBs-q03ZSw7Rc-ev3iy2S-1glM)

Fixes #9220 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9240)
<!-- Reviewable:end -->

ianvernon commented on daemon/endpoint.go:
```diff
@@ -54,10 +54,11 @@ func (h *getEndpoint) Handle(params GetEndpointParams) middleware.Responder {
 }
 
 func (d *Daemon) getEndpointList(params GetEndpointParams) []*models.Endpoint {
+	maxGoroutines := 10
```
What is the reasoning behind using '10' here? Was there an empirical measurement done? Before, we have used the number of CPUs.

fristonio replied:
Hey @ianvernon This is just a dummy value and the PR is still in progress. Me and @aanm are working on benchmarking/profiling the `getEndpointList` function to decide upon a value/need of maximum go routines limit for this particular function. We are still experimenting and have not reached upon a conclusion yet, I will update the PR with better description once we are done with our research.

ianvernon replied:
Got it, sorry for being a bit early on the review !

aanm commented on pkg/endpoint/endpoint_test.go:
```diff
@@ -656,6 +656,15 @@ func (s *EndpointSuite) TestEndpointEventQueueDeadlockUponDeletion(c *C) {
+func BenchmarkEndpointGetModel(b *testing.B) {
+	e := NewEndpointWithState(&suite, &FakeEndpointProxy{}, 123, StateCreating)
+
+	b.ReportAllocs()
```
`b.ResetTimer()`

aanm commented on daemon/endpoint.go:
```diff
@@ -69,18 +71,35 @@ func (d *Daemon) getEndpointList(params GetEndpointParams) []*models.Endpoint {
 	}
 
+	// Send the endpoints to be aggregated a models to the endpoint channel.
+	go func(epsChan chan<- *endpoint.Endpoint, eps []*endpoint.Endpoint) {
```
Since `epsChan` is a buffered channel, this doesn't need to be a go routine

fristonio replied:
If I understood this correctly then, `epsModelChan` is also a buffered channel with size maxGoroutines which is used at line 108. If the number of endpoints is greater than 2 * maxGoroutines then both `epsCh` and `epsModelChan` will be filled and the mentioned function not being a goroutine will block and there will be a deadlock situation.

aanm replied:
@fristonio very good point. Now that I've noticed, both `epsCh` and `epModelsCh` don't need to have a length of `maxGoroutines`. Either `epModelsCh := make(chan *models.Endpoint, len(eps))` and we can remove this go routine or have the go routine and leave the `epModelsCh := make(chan *models.Endpoint, maxGoroutines)`. Either way it's good as it is now. :+1:

tgraf commented on daemon/endpoint.go:
```diff
@@ -57,7 +58,7 @@ func (h *getEndpoint) Handle(params GetEndpointParams) middleware.Responder {
 
 func (d *Daemon) getEndpointList(params GetEndpointParams) []*models.Endpoint {
-	maxGoroutines := 10
+	maxGoroutines := runtime.NumCPU()
```
Can you merge these changes into the first commit? I was confused to still see `:= 10` while the commit description talked about max number of cores.

fristonio replied:
Yeah sure, I have updated the pull request.

tgraf reviewed (changes requested):
Awesome work! Thanks for putting the profiling data into the PR as well, it really helped me understand the rational.

fristonio committed:
daemon: split endpoint list process to use go routines workerpool

* Currently when doing `cilium endpoint list` a number of goroutines equal
to the number of endpoints are created which then extract the current
state of individual endpoints. This does not make much sense since we
don't have any blocking work when getting model for an endpoint using
`endpoint.GetModel`.

* This commit creates a pool of goroutines with a limit on the
maximum number of goroutines(equal to the number of CPUs) and then
distribute this work among those.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
endpoint: add endpoint get model function benchmarking test

* Add benchmarking tests for endpoint `GetModel` function to benchmark
various function which corresponds to getEndpoints* functions.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

aanm reviewed (changes requested):
@fristonio it seems the unit tests are failing related with your changes. I would test out if the code works with `maxGoroutines := 1` trying to figure out what the problem is and then set a minimal limit of 2 go routines.
